Podman Image SCP rootful to rootless transfer

cdoern opened the pull request:
Added functionality for users to transfer images from root storage to rootless storage without using sshd. This is
done through rootful podman by running `sudo podman image scp root@localhost::image user@localhost:: the user is needed
in order to find and use their uid/gid to exec a new process.

added necessary tests, and functions for this implementation. Created new image function `Transfer` so that
the underlying code is majorly removed from CLI.

Signed-off-by: cdoern <cdoern@redhat.com>

TomSweeneyRedHat commented on docs/source/markdown/podman-image-scp.1.md:
````diff
@@ -62,6 +70,16 @@ Storing signatures
 ```
 
+```
+$ sudo podman image scp --rootless --user=charliedoern alpine
````
```suggestion
$ sudo podman image scp --rootless --user=myusername alpine
```

TomSweeneyRedHat commented on docs/source/markdown/podman-image-scp.1.md:
```diff
@@ -28,6 +28,14 @@ Suppress the output
 
+#### **--rootless**
+
+Rootless determines whether we are loading locally from root storage to rootless storage. This flag must be used in conjunction with --user.
```
loading what?  This doesn't really tell the user what this accomplishes.  How's about:

```suggestion
The rootless option allows a user to transfer images from the root storage to the user's 'rootles' storage.  This option must be used in conjunction with the --user option.
```

TomSweeneyRedHat commented on pkg/domain/infra/abi/images.go:
```diff
@@ -330,6 +336,80 @@ func (ir *ImageEngine) Push(ctx context.Context, source string, destination stri
+// Transfer moves images from root to rootless storage so the user specified in the scp call can access and use the image modified by root
+func (ir *ImageEngine) Transfer(ctx context.Context, scpOpts entities.ImageScpOptions) error {
+	if scpOpts.User == "" {
+		return errors.Wrapf(define.ErrInvalidArg, "you must define a user when transferring from root to rootless storage")
```
Over/under on how long it takes someone to ask to transfer from rootless to root?

TomSweeneyRedHat commented on pkg/domain/infra/tunnel/images.go:
```diff
@@ -122,6 +123,10 @@ func (ir *ImageEngine) Pull(ctx context.Context, rawImage string, opts entities.
 }
 
+func (ir *ImageEngine) Transfer(ctx context.Context, scpOpts entities.ImageScpOptions) error {
+	return errors.Wrapf(define.ErrNotImplemented, "cannot use the remote client to transfer images between root and rootless storage")
```
I like this restriction, but I suspect we might get push back on it.  THis probably ought to be added to a man page somewhere if not already there.

cdoern commented on pkg/domain/infra/abi/images.go:
```diff
@@ -330,6 +336,80 @@ func (ir *ImageEngine) Push(ctx context.Context, source string, destination stri
+	}
+	cred := &syscall.Credential{Uid: 0, Gid: 0, Groups: []uint32{}, NoSetGroups: false}
+	sys := &syscall.SysProcAttr{Credential: cred, Setsid: true, Noctty: false, GidMappingsEnableSetgroups: true, Cloneflags: syscall.CLONE_NEWNS |
+		syscall.CLONE_NEWUTS |
+		syscall.CLONE_NEWIPC |
+		syscall.CLONE_NEWPID |
+		syscall.CLONE_NEWNET |
```
I can try it without these, I thought fully separating the process would make things work smoother.

rhatdan commented:
Why aren't you using the localhost syntax?
podman image scp root@localhost:alpine .
podman image scp alpine root@localhost

None of these fancy flags.

It should be the same syntax, we just need to look for the key word localhost and attempt to shortcut it.

cdoern commented:
@rhatdan I can convert to that syntax with the inclusion of a `user@localhost` to send the image to since we have to execute the process as `sudo podman image scp user@localhost` I was having trouble starting the process rootless and sending the image to root. I will continue to modify it, I'm going to convert this to a draft and see what I can change.

cdoern commented:
@Luap99 I think I began to address some of your concerns with environmental variables, I changed it so that we actually retrieve the users  environmental vars and check for their run directory, using that for the other variables as necessary. @rhatdan I switched the syntax to `sudo podman image scp root@localhost::image user@localhost::` since we need to exec the original podman command through sudo.

cevich commented:
@cdoern I just re-ran all the failed tests.  I see a lot of common flakes scattered about, along with some other definitely strange failures.  Let's see what shakes out in the re-run.  I am in agreement with your theory though, this PR seems to be changing timings enough that it's exposing a bunch of unrelated code/test problems.  Unfortunately I don't think there's any quick fix, since these issues are there to start with.  But I'll stay with it and help you through.

cdoern commented:
Thanks @cevich I really appreciate your help, looks like the ubuntu test made it through and the two remote fedora tests got stuck this time around. Seems to just be luck of the draw what fails and what does not

cevich commented:
> Seems to just be luck of the draw what fails and what does not

Kid of, more like we have a lot of tests, a big matrix, and so there are bugs/race-conditions.  So I think the changes here perturbed things enough for us to notice we've got problems that need fixing.  If it wasn't your PR, it would have been somebody else's.

Looking at the Ubuntu failures now...

cevich commented:
...oof! Looks like more examples of https://github.com/containers/podman/issues/12155 but in Fedora this time.  I noted down log-links in the issue, so I'll just re-run those tasks again.

cdoern commented:
@rhatdan @Luap99 @mheon PTAL. The flaking issue is not resolved @cevich but the tests have passed which coivnces me I didn't do anything wrong other than expose some testing suite unsolvable problems! This is relatively finished, I removed the need for a new user NS. and most chowning of files that should not be chowned. I would like to get this in as the image scp blog discusses it heavily.

cevich commented:
> I didn't do anything wrong other than expose some testing suite unsolvable problems

I agree with this sentiment...so much so, maybe I can use it as an excuse for not paying my taxes? :grin:

Luap99 commented on pkg/domain/infra/abi/images.go:
```diff
@@ -330,6 +334,57 @@ func (ir *ImageEngine) Push(ctx context.Context, source string, destination stri
+		return err
+	}
+
+	cmd := exec.Command("machinectl", "shell", u.Username+"@.host", os.Args[0], "--cgroup-manager=cgroupfs", "load", "--input", scpOpts.Save.Output) // load the new image to the rootless storage
```
I recommend to use exec.LookPath to check if machinectl is there

Luap99 replied:
Also you have to use os.Executable() instead of os.Args[0] because the path must be absolute.

Luap99 replied:
Lastly, please parse the current log level to the new process with "--log-level", logrus.GetLevel().String()

cdoern replied:
fixed

mheon commented on pkg/domain/infra/abi/images.go:
```diff
@@ -330,6 +334,66 @@ func (ir *ImageEngine) Push(ctx context.Context, source string, destination stri
+	}
+	machinectl, err := exec.LookPath("machinectl")
+	if err != nil {
+		cmd := exec.Command("su", u.Username, "--command", podman+" --log-level="+logrus.GetLevel().String()+" --cgroup-manager=cgroupfs load --input="+scpOpts.Save.Output) // load the new image to the rootless storage
```
IMO we should print a warning here (at least on systems w/o systemd) given that `su` + podman is potentially going to fail with a strange error if we have no user session.

cdoern replied:
okay, so just a warning like `defaulting to su since machinectl is not available, su will fail if no user session is available` ?

mheon replied:
Yep, that should be fine

mheon commented:
LGTM

Luap99 commented on pkg/domain/infra/abi/images.go:
```diff
@@ -330,6 +334,67 @@ func (ir *ImageEngine) Push(ctx context.Context, source string, destination stri
+	machinectl, err := exec.LookPath("machinectl")
+	if err != nil {
+		logrus.Warn("defaulting to su since machinectl is not available, su will fail if no user session is available")
+		cmd := exec.Command("su", u.Username, "--command", podman+" --log-level="+logrus.GetLevel().String()+" --cgroup-manager=cgroupfs load --input="+scpOpts.Save.Output) // load the new image to the rootless storage
```
Sorry I overlooked this the last time but you also have to set `--login`

cdoern replied:
ok, just `su -l u.Username...` with everything else the same?

cdoern replied:
fixed

cdoern commented:
wait, machinectl is causing some issues with downloading the image now. Need to investigate after class today.

cdoern commented:
nevermind @Luap99 the issue was with my laptop, tested it on my other machine and works perfectly with machinectl. I think this is good now after solving that.

Luap99 reviewed (approved):
LGTM

Luap99 commented:
/lgtm

rhatdan commented:
@cdoern Not sure what I am doing wrong, but this is definitely broken.

```
 $ ./bin/podman image scp myimage root@localhost:
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
github.com/containers/podman/v3/cmd/podman/images.parseArgs(0xc0003bd9a0, 0x2, 0x2, 0xc0000e4000, 0x1f471b0, 0xc000010718, 0x0)
	/home/dwalsh/podman/cmd/podman/images/scp.go:291 +0x1465
github.com/containers/podman/v3/cmd/podman/images.scp(0x2983040, 0xc0003bd9a0, 0x2, 0x2, 0x0, 0x0)
	/home/dwalsh/podman/cmd/podman/images/scp.go:91 +0x236
github.com/spf13/cobra.(*Command).execute(0x2983040, 0xc000138030, 0x2, 0x2, 0x2983040, 0xc000138030)
	/home/dwalsh/podman/vendor/github.com/spf13/cobra/command.go:856 +0x472
github.com/spf13/cobra.(*Command).ExecuteC(0x2992cc0, 0xc000130030, 0x19ef400, 0x2a55f10)
	/home/dwalsh/podman/vendor/github.com/spf13/cobra/command.go:974 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
	/home/dwalsh/podman/vendor/github.com/spf13/cobra/command.go:902
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/home/dwalsh/podman/vendor/github.com/spf13/cobra/command.go:895
main.Execute()
	/home/dwalsh/podman/cmd/podman/root.go:91 +0xe7
main.main()
	/home/dwalsh/podman/cmd/podman/main.go:39 +0x92
```

cdoern commented:
> @cdoern Not sure what I am doing wrong, but this is definitely broken.
> 
> ```
>  $ ./bin/podman image scp myimage root@localhost:
> panic: runtime error: index out of range [1] with length 1
> 
> goroutine 1 [running]:
> github.com/containers/podman/v3/cmd/podman/images.parseArgs(0xc0003bd9a0, 0x2, 0x2, 0xc0000e4000, 0x1f471b0, 0xc000010718, 0x0)
> 	/home/dwalsh/podman/cmd/podman/images/scp.go:291 +0x1465
> github.com/containers/podman/v3/cmd/podman/images.scp(0x2983040, 0xc0003bd9a0, 0x2, 0x2, 0x0, 0x0)
> 	/home/dwalsh/podman/cmd/podman/images/scp.go:91 +0x236
> github.com/spf13/cobra.(*Command).execute(0x2983040, 0xc000138030, 0x2, 0x2, 0x2983040, 0xc000138030)
> 	/home/dwalsh/podman/vendor/github.com/spf13/cobra/command.go:856 +0x472
> github.com/spf13/cobra.(*Command).ExecuteC(0x2992cc0, 0xc000130030, 0x19ef400, 0x2a55f10)
> 	/home/dwalsh/podman/vendor/github.com/spf13/cobra/command.go:974 +0x375
> github.com/spf13/cobra.(*Command).Execute(...)
> 	/home/dwalsh/podman/vendor/github.com/spf13/cobra/command.go:902
> github.com/spf13/cobra.(*Command).ExecuteContext(...)
> 	/home/dwalsh/podman/vendor/github.com/spf13/cobra/command.go:895
> main.Execute()
> 	/home/dwalsh/podman/cmd/podman/root.go:91 +0xe7
> main.main()
> 	/home/dwalsh/podman/cmd/podman/main.go:39 +0x92
> ```

@rhatdan you need to specify a user. I will submit a patch to not allow for just one argument with root. the syntax is `sudo podman image scp root@localhost::image user@localhost::` this is necessary to chown the image file to a specific user. I guess i could set it up so if no user is specified it defaults to 1000:1000

rhatdan commented:
User should be assumed to be me, if I don't specify.

podman image scp myimage root@localhost::
podman image scp root@localhost::foo bar

Should be ok syntax.

cdoern commented:
@rhatdan the issue with assuming to be "me" is that this command (root to rootless) needs to be run using sudo. Is there a way to get default user or should i just assume 1000:1000.

Luap99 commented:
@cdoern You can use `SUDO_USER` and `SUDO_UID` env vars when the command is run with `sudo`. Otherwise I would just error and say that you need to specify the user.

cdoern commented:
> @cdoern You can use `SUDO_USER` and `SUDO_UID` env vars when the command is run with `sudo`. Otherwise I would just error and say that you need to specify the user.

@Luap99 yep, I just submitted a PR using SUDO_USER

rhatdan commented:
Again, I do not want to run as sudo, I want magic to happen.  If I am running with sudo, then fine.  Looking at SUDO_USER would be fine.

My question why can't 
podman image scp myimage root@localhost::
Work?  I would figure this would just do

podman image save $XDG_RUNTIME_DIR:/myimage; sudo podman load $XDG_RUNTIME_DIR:/myimage

Under the covers.

Luap99 commented:
> Again, I do not want to run as sudo, I want magic to happen. If I am running with sudo, then fine. Looking at SUDO_USER would be fine.
> 
> My question why can't podman image scp myimage root@localhost:: Work? I would figure this would just do
> 
> podman image save $XDG_RUNTIME_DIR:/myimage; sudo podman load $XDG_RUNTIME_DIR:/myimage Under the covers.
> 
> sudo podman image scp root@localhost::foobar .
> 
> Would do sudo podman image save $XDG_RUNTIME_DIR:/foobar; podman load $XDG_RUNTIME_DIR:/foobar

This does not work because you cannot run sudo inside the podman user namespace.

cdoern commented:
see [comment on new PR](https://github.com/containers/podman/pull/12224#issuecomment-963512749) it seems to be impossible to invoke sudo inside of a rootless podman user NS

rhatdan commented:
Seems to me that we should be able to break into a couple of different threads, one that enters the user namespace to do its thing and one to do the sudo command and exec itself.

cdoern commented:
correct which command? the original `podman image scp` or the exec'd `load`? If this is to be done rootless, the save command must also be an exec.Command reaching into root storage.

giuseppe commented:
> Seems to me that we should be able to break into a couple of different threads, one that enters the user namespace to do its thing and one to do the sudo command and exec itself.

a thread cannot enter a user namespace.  The entire process must do it (when it is single threaded).

rhatdan commented:
This all be done with two execs, One into the user namespace and one with sudo?
podman image scp alpine root@localhost
---> exec podman image save alpine XDG_RUNTIME_DIR/alpine
---> exec sudo podman image load XDG_RUNTIME_DIR/alpine
